fix-290

hexfaker opened the pull request:
Fix #290 feat. @bagxi

hexfaker committed:
Add `batch2device()` call to `predict_batch()`

bagxi committed:
[fix] bin/check_dl.sh CUDA errors (#322)

* move loaded from previous stage optimizer to available device

hexfaker committed:
Small changes in `check_dl.sh`

Scitator commented on catalyst/dl/runner/supervised.py:
```diff
@@ -83,7 +83,7 @@ def _process_output_none(self, output: Mapping[str, Any]):
 
     def predict_batch(self, batch: Mapping[str, Any]):
-        output = self._process_input(batch)
+        output = self._process_input(self._batch2device(batch, self.device))
```
why do we need it?
this already done in https://github.com/catalyst-team/catalyst/blob/c2be013a6884fb55c5fa343dbbd7dc967d2ca0b5/catalyst/dl/core/runner.py#L102

Scitator replied:
or you propose to use this method for per batch inference? 🤔

hexfaker replied:
> or you  propose to use this method for per batch inference?

Exactly!

Scitator replied:
could you please use:
```
batch = self._batch2device(batch, self.device)
output = self._process_input(batch)
```
it would be a little bit easier to read

Scitator replied:
or, as even better solution, could you add `_predict_batch` method, that would do only model forward, use it in `Runner. _run_batch`
and add additional `predict_batch` method based on `_batch2device` and `_predict_batch`

this should remove duplication of `_batch2device` usage during training (which is more important, I guess)

Scitator commented on bin/check_dl.sh:
```diff
@@ -1,11 +1,11 @@
 tar -xf ./data/isbi.tar.gz -C ./data/
-
-(set -e; for f in examples/_tests_scripts/*.py; do PYTHONPATH=./catalyst:${PYTHONPATH} python "$f"; done)
+set -e
```
could you please place it on top of the script?

hexfaker committed:
Introduce new method `forward()`

Method introduced in order to more clearly separate jupyter-api of
SupervisedRunner and model call in all other Runners and fix bug related
to device move when `predict_batch()` called manually from outside of runner

Scitator commented on bin/check_dl.sh:
```diff
@@ -2,9 +2,9 @@
+wget -P data -N https://www.dropbox.com/s/0rvuae4mj6jn922/isbi.tar.gz
 tar -xf ./data/isbi.tar.gz -C ./data/
-(set -e; for f in examples/_tests_scripts/{dl,z}*.py; do PYTHONPATH=./catalyst:${PYTHONPATH} python "$f"; done)
+(set -e; for f in examples/_tests_scripts/{dl,z_dl}*.py; do PYTHONPATH=./catalyst:${PYTHONPATH} python "$f"; done)
```
Sorry, but we have to check all our integration scripts, both DL and RL.

hexfaker replied:
Whatever, but the naming is unclear then and dl tests require dependencies beyond requirements. 
There are separrate CI stages for rl, why rl and dl test are still mixed?

Scitator replied:
cause it's much easier and faster to check them both in one place

Scitator replied:
you can rename script to `check_dl_rl_scripts`, nevertheless we need all test checks

hexfaker replied:
Okay, but this should probably be done in another PR, not by me. I reverted my changes related to this issue.

Scitator commented on catalyst/dl/core/runner.py:
```diff
@@ -107,7 +107,7 @@ def _run_batch(self, batch):
         self._run_event("batch_start")
         self.state.timer.start("_timers/model_time")
-        self.state.output = self.predict_batch(batch)
+        self.state.output = self.forward(batch)
```
you still need to pass batch to device, so `predict_batch` required

hexfaker replied:
Nope, batch is already moved to device few lines above in the same function

Scitator replied:
okay, i see
than, could you please move `forward` method on top of `predict_batch`?

like
```
def forward(...)
   ...

def predict_batch(...)
    ...
    self.forward()

def _run_batch(...):
    ...
    self.forward()
```

hexfaker committed:
Polish
